subsys: fs: fcb: struct flash_sector usually const

JPHutchins opened the pull request:
Generally, the flash_sector array is known at compile time and is immutable, therefore developers may like to store it in ROM. The only function that requires a mutable struct flash_sector seems to be flash_area_get_sectors(), allowing the developer to store a mutable sector definition in persistent RAM rather than ROM.  An example of this approach is in setting_fcb.c.

JPHutchins commented:
I am able to reproduce the CI error on local:

```
./scripts/coccicheck --mode=report --cocci=scripts/coccinelle/same_identifier.cocci | grep fcb
/home/jp/repos/zephyrproject/zephyr/tests/subsys/fs/fcb/src/fcb_test_last_of_n.c:13:13-16: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/subsys/fs/fcb/fcb_append.c:59:23-26: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/subsys/fs/fcb/fcb_append.c:120:30-33: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/subsys/fs/fcb/fcb_append.c:39:34-37: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/subsys/fs/fcb/fcb_append.c:14:61-64: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/tests/subsys/fs/fcb/src/fcb_test_empty_walk.c:13:13-16: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/tests/subsys/fs/fcb/src/fcb_test_append_too_big.c:12:14-17: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/tests/subsys/fs/fcb/src/fcb_test_rotate.c:12:13-16: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/tests/subsys/fs/fcb/src/fcb_test_multiple_scratch.c:12:13-16: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/tests/subsys/fs/fcb/src/fcb_test_append_fill.c:12:13-16: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/tests/subsys/fs/fcb/src/fcb_test_init.c:13:13-16: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/subsys/fs/fcb/fcb_walk.c:15:25-28: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/tests/subsys/fs/fcb/src/fcb_test_reset.c:12:13-16: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/subsys/fs/fcb/fcb_rotate.c:12:23-26: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/subsys/fs/fcb/fcb_getnext.c:108:24-27: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/subsys/fs/fcb/fcb_getnext.c:14:34-37: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/subsys/fs/fcb/fcb_getnext.c:43:31-34: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
/home/jp/repos/zephyrproject/zephyr/subsys/fs/fcb/fcb_getnext.c:33:58-61: WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
```

But not sure how to improve!

JPHutchins commented:
Needs rebase.

de-nordic reviewed (changes requested):
Please keep the const type qualifier  between type definition and after storage class (extern).
The change sticks out from all the other code, including modified files:
![image](https://github.com/user-attachments/assets/44dc5c4c-999c-419f-b7d8-e907777067fd)

kartben commented:
> Needs rebase.

Yes but you're the author? 🙂

JPHutchins committed:
fs: fcb: struct flash_sector usually const

Generally, the flash_sector array is known at compile time and
is immutable, therefore developers may like to store it in ROM.
The only function that requires a mutable struct flash_sector
seems to be flash_area_get_sectors(), allowing the developer
to store a mutable sector definition in persistent RAM rather
than ROM.  An example of this approach is in setting_fcb.c.

Signed-off-by: JP Hutchins <jp@intercreate.io>

JPHutchins commented:
Any insight on these compliance check problems? They are present on lines that are not modified by this PR.

```
Error: subsys/fs/fcb/fcb_append.c:14:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
subsys/fs/fcb/fcb_getnext.c:33:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
subsys/fs/fcb/fcb_walk.c:15:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
Error: Process completed with exit code 1.
```

de-nordic commented:
> Any insight on these compliance check problems? They are present on lines that are not modified by this PR.
> 
> ```
> Error: subsys/fs/fcb/fcb_append.c:14:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
> subsys/fs/fcb/fcb_getnext.c:33:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
> subsys/fs/fcb/fcb_walk.c:15:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
> Error: Process completed with exit code 1.
> ```

I guess CI does not like name of variable to be the same as type used to define it.
It was that before, but you have touched the line, so CI can now insist on you to fix it.

kartben commented:
ya most of these files hadn't been touched in many years, and the coding guidelines check has been introduced since then. Luckily the fix should be pretty straightforward

fabiobaltieri commented on tests/subsys/fs/fcb/src/main.c:
```diff
@@ -30,25 +30,11 @@ uint8_t fcb_test_erase_value;
-	}
-};
-
+const struct flash_sector test_fcb_sector[] = {
+	[0] = {.fs_off = 0, .fs_size = SECTOR_SIZE},
+	[1] = {.fs_off = SECTOR_SIZE, .fs_size = SECTOR_SIZE},
+	[2] = {.fs_off = 2 * SECTOR_SIZE, .fs_size = SECTOR_SIZE},
+	[3] = {.fs_off = 3 * SECTOR_SIZE, .fs_size = SECTOR_SIZE}};
```
unrelated change? same in another couple of files, PRs should  nto change unrelated code

de-nordic replied:
@JPHutchins If you did run clang format on the code, please do not do that. At least not with other changes.

de-nordic commented:
Any movement here?
